src/DataObject/Service/ExecutionEngine/CloneService.php

```diff
@@ -0,0 +1,179 @@
+use Pimcore\Bundle\StudioBackendBundle\Exception\Api\NotFoundException;
+use Pimcore\Bundle\StudioBackendBundle\Exception\Api\UserNotFoundException;
+use Pimcore\Bundle\StudioBackendBundle\ExecutionEngine\Util\Config;
+use Pimcore\Bundle\StudioBackendBundle\ExecutionEngine\Util\EnvironmentVariables;
+use Pimcore\Bundle\StudioBackendBundle\ExecutionEngine\Util\Jobs;
+use Pimcore\Bundle\StudioBackendBundle\Security\Service\SecurityServiceInterface;
+use Pimcore\Bundle\StudioBackendBundle\Util\Constants\ElementPermissions;
+use Pimcore\Model\DataObject;
+use Pimcore\Model\DataObject\Service as DataObjectService;
+use Pimcore\Model\UserInterface;
+use function sprintf;
+
+/**
+ * @internal
+ */
+final readonly class CloneService implements CloneServiceInterface
+{
+    public function __construct(
+        private DataObjectServiceInterface $dataObjectService,
+        private DataObjectSearchServiceInterface $dataObjectSearchService,
+        private JobExecutionAgentInterface $jobExecutionAgent,
+        private SecurityServiceInterface $securityService,
+        private SynchronousProcessingServiceInterface $synchronousProcessingService
+    ) {
+    }
+
+    /**
+     * @throws AccessDeniedException
+     * @throws ElementSavingFailedException
+     * @throws ForbiddenException
+     * @throws NotFoundException
+     * @throws UserNotFoundException
+     */
+    public function cloneDataObjects(
+        int $sourceId,
+        int $parentId,
+        CloneParameters $parameters,
+    ): ?int {
+        $user = $this->securityService->getCurrentUser();
+        $source = $this->dataObjectService->getDataObjectElement(
+            $user,
+            $sourceId,
+        );
+        $newParent = $this->cloneDataObject(
+            $source,
+            $this->dataObjectService->getDataObjectElement(
+                $user,
+                $parentId,
+            ),
+            $this->securityService->getCurrentUser()
+        );
+
+        if (!$parameters->isRecursive() || !$source->hasChildren()) {
+            return null;
+        }
+
+        return $this->cloneChildrenWithExecutionEngine(
+            user: $user,
+            originalParent: $source,
+            newParent: $newParent,
+            updateReferences: $parameters->isUpdateReferences(),
+        );
+
+    }
+
+    /**
+     * @throws ElementSavingFailedException|ForbiddenException
+     */
+    public function cloneDataObject(
+        DataObject $source,
+        DataObject $parent,
+        UserInterface $user
+    ): DataObject {
+        if (!$parent->isAllowed(ElementPermissions::CREATE_PERMISSION)) {
+            throw new ForbiddenException(
+                sprintf(
+                    'Missing permissions on target element %s',
+                    $parent->getId()
+                )
+            );
+        }
+
+        $this->synchronousProcessingService->enable();
+
+        $dataObject = (new DataObjectService())->copyAsChild(
```

Comment:
Can we use the resolver here?

Reply:
Not sure how since this is not a registered service

Reply:
Core is somtimes mind boggling :D